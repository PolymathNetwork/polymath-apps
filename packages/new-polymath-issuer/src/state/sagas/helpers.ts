import { fork, take, call, Pattern, HelperFunc0 } from 'redux-saga/effects';
import { Action } from 'redux';

export const takeOneAtATime = <A extends Action>(
  patternOrChannel: Pattern,
  saga: HelperFunc0<A>
) =>
  fork(function*() {
    while (true) {
      const action: A = yield take(patternOrChannel);
      yield call(saga, action);
    }
  });
