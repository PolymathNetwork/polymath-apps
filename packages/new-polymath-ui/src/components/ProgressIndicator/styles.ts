import { styled } from '~/styles';
import { ulReset } from '~/styles/utils';

export const Container = styled.ul<{ vertical: boolean }>`
  ${ulReset};
  display: grid;
  grid-auto-rows: 1fr;
  grid-auto-columns: 1fr;
  list-style: none;
  font-size: ${({ theme }) => theme.fontSizes.baseText};

  ${({ vertical }) =>
    vertical
      ? `
      grid-auto-flow: row;
      height: 100%;
      justify-content: space-between;
      `
      : `
      grid-auto-flow: column;
      height: auto;
    `};
`;
