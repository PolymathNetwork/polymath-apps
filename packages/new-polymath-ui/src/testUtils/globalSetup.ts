import { cleanup } from 'react-testing-library';

afterEach(cleanup);
