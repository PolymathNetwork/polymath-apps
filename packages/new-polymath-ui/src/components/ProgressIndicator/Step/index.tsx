import React, { FC } from 'react';
import { Paragraph } from '~/components/Paragraph';
import { SvgCheckmark } from '~/images/icons/Checkmark';
import * as sc from './styles';
import { iconsSize } from './styles';

export interface StepProps {
  /**
   * Specify whether the step is the current step
   */
  isCurrent?: boolean;
  /**
   * Specify whether the step has been completed
   */
  isComplete?: boolean;
  /**
   * Provide the label for the <ProgressStep>
   */
  label?: string;
  isVertical?: boolean;
  isOrdered?: boolean;
  index?: number;
}

export const Step: FC<StepProps> = ({
  label,
  index,
  isCurrent,
  isComplete,
  isVertical,
  isOrdered,
  ...otherProps
}) => {
  const stepNumber = (index || 0) + 1;

  const NumberedSvg: FC = () => (
    <Paragraph as="span" fontWeight="bold" fontSize="inherit">
      {stepNumber}
    </Paragraph>
  );

  const CurrentStepSvg: FC<React.SVGAttributes<SVGElement>> = props => (
    <svg role="img" viewBox="0 0 24 24" {...props}>
      <title>Current step</title>
      <circle fill="currentColor" cx="12" cy="12" r="8" />
    </svg>
  );
  const IncompleteStepSvg: FC<React.SVGAttributes<SVGElement>> = (
    props: React.SVGAttributes<SVGElement>
  ) => (
    <svg
      role="img"
      viewBox="0 0 24 24"
      data-testid="incomplete-step-svg"
      {...props}
    >
      <title>Incomplete step</title>
      <circle fill="currentColor" cx="12" cy="12" r="8" />
    </svg>
  );

  const CompleteSvg = isComplete && SvgCheckmark;
  const CurrentSvg = isCurrent && (isOrdered ? NumberedSvg : CurrentStepSvg);
  const IncompleteSvg =
    !isComplete && isOrdered ? NumberedSvg : IncompleteStepSvg;

  return (
    <sc.Container
      isCurrent={isCurrent}
      isComplete={isComplete}
      isVertical={isVertical}
      {...otherProps}
    >
      <sc.Inner>
        <sc.Icon
          Asset={CurrentSvg || CompleteSvg || IncompleteSvg}
          width={iconsSize}
          height={iconsSize}
          scale={0.8}
        />
        {!!label && (
          <sc.Label
            mb={0}
            mt={isVertical ? 0 : `calc(${iconsSize} / 2)`}
            ml={isVertical ? `calc(${iconsSize} / 2)` : 0}
            fontWeight="bold"
            fontSize="inherit"
          >
            {label}
          </sc.Label>
        )}
      </sc.Inner>
      <sc.ProgressLine />
    </sc.Container>
  );
};
