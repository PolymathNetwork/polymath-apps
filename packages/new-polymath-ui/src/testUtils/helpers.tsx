import React, { Fragment } from 'react';
import { render } from 'react-testing-library';
import { ThemeProvider } from '~/styles';

const customRender = (node: React.ReactElement<any>, ...opts: any[]) => {
  const { rerender, ...result } = render(
    <ThemeProvider>{node}</ThemeProvider>,
    ...opts
  );

  const customRerender = (rerenderNode: React.ReactElement<any>) => {
    return rerender(<Fragment>{rerenderNode}</Fragment>);
  };

  return { rerender: customRerender, ...result };
};

export * from 'react-testing-library';
export { customRender as render };
