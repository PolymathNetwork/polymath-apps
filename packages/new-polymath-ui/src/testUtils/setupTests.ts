import 'babel-polyfill';
import 'jest-dom/extend-expect';
import 'jest-styled-components';
import { cleanup } from '~/testUtils/helpers';

// We mock this module since it cannot be parsed and is ignored by jests transformer for some reason
jest.mock('typeface-overpass', () => ({}));
jest.mock('simplebar/dist/simplebar.css', () => ({}));

// Unmounts React trees that were mounted with render.
afterEach(cleanup);
