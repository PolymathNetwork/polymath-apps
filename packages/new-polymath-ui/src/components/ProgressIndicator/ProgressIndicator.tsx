import React from 'react';
import { Step } from './Step';
import * as sc from './styles';

interface Props {
  /**
   * Provide <Step> components to be rendered in the
   * <ProgressIndicator>
   */
  children?: React.ReactNode;
  /**
   * Specify the current step array index
   */
  currentIndex: number;
  vertical: boolean;
  ordered?: boolean;
}

export class ProgressIndicator extends React.Component<Props> {
  public static Step = Step;

  public renderSteps = () => {
    const { vertical, ordered } = this.props;

    return React.Children.map(this.props.children, (child, index) => {
      const childProps = {
        index,
        isVertical: vertical,
        isOrdered: ordered,
        isCurrent: false,
        isComplete: false,
      };

      if (!React.isValidElement(child)) {
        return null;
      }

      if (index === this.props.currentIndex) {
        childProps.isCurrent = true;
      } else if (index < this.props.currentIndex) {
        childProps.isComplete = true;
      }
      return React.cloneElement(child, childProps);
    });
  };

  public render() {
    const { currentIndex, ...otherProps } = this.props;
    return <sc.Container {...otherProps}>{this.renderSteps()}</sc.Container>;
  }
}
