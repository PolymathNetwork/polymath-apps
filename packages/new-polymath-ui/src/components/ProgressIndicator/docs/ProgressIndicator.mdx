---
name: ProgressIndicator
menu: Components
---

import { Playground, PropsTable } from 'docz';
import { Box } from '~/components/Box';
import { ProgressIndicator } from '../ProgressIndicator';
import { StyledProgressIndicator } from './styles';

# ProgressIndicator

<PropsTable of={ProgressIndicator} />

## Basic usage

<Playground>
  <ProgressIndicator currentIndex={2}>
    <ProgressIndicator.Step label="Label 1" />
    <ProgressIndicator.Step label="Label 2" />
    <ProgressIndicator.Step label="Label 3" />
    <ProgressIndicator.Step label="Label 4" />
  </ProgressIndicator>
</Playground>

## Vertical

<Playground>
  <Box height={400}>
    <ProgressIndicator currentIndex={2} vertical ordered>
      <ProgressIndicator.Step label="Label 1" />
      <ProgressIndicator.Step label="Label 2" />
      <ProgressIndicator.Step label="Label 3" />
      <ProgressIndicator.Step label="Label 4" />
    </ProgressIndicator>
  </Box>
</Playground>

## Styled

<Playground>
  <Box height={400}>
    <StyledProgressIndicator currentIndex={2} vertical ordered>
      <ProgressIndicator.Step label="Label 1" />
      <ProgressIndicator.Step label="Label 2" />
      <ProgressIndicator.Step label="Label 3" />
      <ProgressIndicator.Step label="Label 4" />
    </StyledProgressIndicator>
  </Box>
</Playground>
