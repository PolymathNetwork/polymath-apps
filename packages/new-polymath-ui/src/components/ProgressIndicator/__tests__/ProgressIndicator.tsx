import { within } from 'react-testing-library';
import React from 'react';
import { ProgressIndicator } from '../ProgressIndicator';
import { render } from '~/testUtils/helpers';

describe('ProgressIndicator', () => {
  test('renders without crashing', () => {
    const { container } = render(
      <ProgressIndicator vertical currentIndex={0}>
        <ProgressIndicator.Step label="foo" />
      </ProgressIndicator>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('ordered', () => {
    test('correctly numbers the steps', () => {
      const { getByTestId } = render(
        <ProgressIndicator vertical currentIndex={1} ordered>
          <ProgressIndicator.Step label="First Step" data-testid="first-step" />
          <ProgressIndicator.Step
            label="Second Step"
            data-testid="second-step"
          />
          <ProgressIndicator.Step label="Third Step" data-testid="third-step" />
        </ProgressIndicator>
      );

      const firstStep = getByTestId('first-step');
      const secondStep = getByTestId('second-step');
      const thirdStep = getByTestId('third-step');

      // Since it is a completed step it should not show its number
      expect(firstStep).not.toHaveTextContent('1');
      expect(secondStep).toHaveTextContent('2');
      expect(thirdStep).toHaveTextContent('3');
    });
  });

  test('marks next steps as incomplete', () => {
    const { getByTestId } = render(
      <ProgressIndicator vertical currentIndex={0}>
        <ProgressIndicator.Step label="First Step" data-testid="first-step" />
        <ProgressIndicator.Step label="Second Step" data-testid="second-step" />
        <ProgressIndicator.Step label="Third Step" data-testid="third-step" />
      </ProgressIndicator>
    );

    const firstStep = getByTestId('first-step');
    const secondStep = getByTestId('second-step');
    const thirdStep = getByTestId('third-step');

    expect(() => {
      within(firstStep).getByTestId('incomplete-step-svg');
    }).toThrow();
    within(secondStep).getByTestId('incomplete-step-svg');
    within(thirdStep).getByTestId('incomplete-step-svg');
  });
});
