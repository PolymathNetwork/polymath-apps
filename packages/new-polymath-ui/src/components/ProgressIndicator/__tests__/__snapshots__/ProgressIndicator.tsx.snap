// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressIndicator renders without crashing 1`] = `
.c6 {
  position: absolute;
  background-color: currentColor;
}

.c3 {
  display: inline-block;
  vertical-align: middle;
  background-color: transparent;
  width: 1.6em;
  height: 1.6em;
  border-radius: 50%;
  border-style: solid;
  border-width: 2px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c3 svg {
  display: block;
  width: 100%;
  height: 100%;
  padding: 19.999999999999996%;
}

.c4 {
  box-sizing: border-box;
  margin-bottom: 0;
  margin-left: calc(1.6em / 2);
  margin-top: 0;
  font-weight: 400;
  line-height: 1.5;
  font-size: inherit;
  overflow-wrap: break-word;
  line-height: 1.6em;
  color: currentColor;
}

.c4:last-child {
  margin-bottom: 0;
}

.c1 {
  position: relative;
  min-width: 7rem;
  -webkit-transition: 250ms all cubic-bezier(0.5,0,0.1,1);
  transition: 250ms all cubic-bezier(0.5,0,0.1,1);
  overflow: visible;
  color: #252D6B;
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c1:first-child {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

.c1 .c5 {
  left: calc(1.6em / 2);
  top: 1.6em;
  bottom: 0;
  width: 2px;
}

.c1 .c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c1:last-child .c5 {
  display: none;
}

.c0 {
  list-style: none;
  padding: 0;
  margin: 0;
  display: grid;
  grid-auto-rows: 1fr;
  grid-auto-columns: 1fr;
  list-style: none;
  font-size: 0.875rem;
  grid-auto-flow: row;
  height: 100%;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

<ul
  class="c0"
>
  <li
    class="c1"
  >
    <div
      class="c2 "
    >
      <span
        class="c3"
      >
        <svg
          role="img"
          viewBox="0 0 24 24"
        >
          <title>
            Current step
          </title>
          <circle
            cx="12"
            cy="12"
            fill="currentColor"
            r="8"
          />
        </svg>
      </span>
      <p
        class="c4"
        font-size="inherit"
        font-weight="bold"
      >
        foo
      </p>
    </div>
    <span
      class="c5 c6"
    />
  </li>
</ul>
`;
